page.go

```diff
@@ -38,50 +38,338 @@ type FullArtistPage struct {
+	}
+	*p = *newP
+
+	return nil
+}
+
+// PreviousPage sets the page content to the previous page if available
+func (p *SimplePlaylistPage) PreviousPage(c *Client) error {
+	if p.Previous == "" {
+		return ErrNoMorePages
+	}
+
+	newP := &SimplePlaylistPage{}
+	err := c.get(p.Previous, newP)
+	if err != nil {
+		return err
+	}
+	*p = *newP
+
+	return nil
+}
+
 // SimpleTrackPage contains SimpleTracks returned by the Web API.
 type SimpleTrackPage struct {
 	basePage
 	Tracks []SimpleTrack `json:"items"`
 }
 
+// NextPage sets the page content to the next page if availabe
+func (p *SimpleTrackPage) NextPage(c *Client) error {
+	if p.Next == "" {
+		return ErrNoMorePages
+	}
+
+	newP := &SimpleTrackPage{}
+	err := c.get(p.Next, newP)
+	if err != nil {
+		return err
+	}
+	*p = *newP
+
+	return nil
+}
+
+// PreviousPage sets the page content to the previous page if available
+func (p *SimpleTrackPage) PreviousPage(c *Client) error {
+	if p.Previous == "" {
+		return ErrNoMorePages
+	}
+
+	newP := &SimpleTrackPage{}
+	err := c.get(p.Previous, newP)
+	if err != nil {
+		return err
+	}
+	*p = *newP
+
+	return nil
+}
+
 // FullTrackPage contains FullTracks returned by the Web API.
 type FullTrackPage struct {
 	basePage
 	Tracks []FullTrack `json:"items"`
 }
 
+// NextPage sets the page content to the next page if availabe
+func (p *FullTrackPage) NextPage(c *Client) error {
+	if p.Next == "" {
+		return ErrNoMorePages
+	}
+
+	newP := &FullTrackPage{}
+	err := c.get(p.Next, newP)
+	if err != nil {
+		return err
+	}
+	*p = *newP
+
+	return nil
+}
+
+// PreviousPage sets the page content to the previous page if available
+func (p *FullTrackPage) PreviousPage(c *Client) error {
+	if p.Previous == "" {
+		return ErrNoMorePages
+	}
+
+	newP := &FullTrackPage{}
+	err := c.get(p.Previous, newP)
+	if err != nil {
+		return err
+	}
+	*p = *newP
+
+	return nil
+}
+
 // SavedTrackPage contains SavedTracks return by the Web API.
 type SavedTrackPage struct {
 	basePage
 	Tracks []SavedTrack `json:"items"`
 }
 
+// NextPage sets the page content to the next page if availabe
+func (p *SavedTrackPage) NextPage(c *Client) error {
+	if p.Next == "" {
+		return ErrNoMorePages
+	}
+
+	newP := &SavedTrackPage{}
+	err := c.get(p.Next, newP)
+	if err != nil {
+		return err
+	}
+	*p = *newP
+
+	return nil
+}
+
+// PreviousPage sets the page content to the previous page if available
+func (p *SavedTrackPage) PreviousPage(c *Client) error {
+	if p.Previous == "" {
+		return ErrNoMorePages
+	}
+
+	newP := &SavedTrackPage{}
+	err := c.get(p.Previous, newP)
+	if err != nil {
+		return err
+	}
+	*p = *newP
+
+	return nil
+}
+
 // PlaylistTrackPage contains information about tracks in a playlist.
 type PlaylistTrackPage struct {
 	basePage
 	Tracks []PlaylistTrack `json:"items"`
 }
 
+// NextPage sets the page content to the next page if availabe
+func (p *PlaylistTrackPage) NextPage(c *Client) error {
+	if p.Next == "" {
+		return ErrNoMorePages
+	}
+
+	newP := &PlaylistTrackPage{}
```

Comment:
I found writing this way looked way to long for what it needed to do.
I also tried thes following, but it's slower (extra overhead from `defer`)

```
newP := &PlaylistTrackPage{}
defer func() { *p = *newP }()
return c.get(p.Next, newP)
```

From my accurate testing on a wifi hotspot in my local dance school I got on average 350ms (lowest 349ms) getting all pages (playlist with 150 items so 2 pages) using the defer way, using the way I left at the pr at this point, it took on average 325ms (lowest 318ms).

May not be very much but I'd take this gain.

Reply:
I like it without the defer, but not for speed reasons (the overhead of a defer is negligible compared to the round-trip time of the request).  I just find it easier to read without the defer.